### ISSUE TITLE
RTL text with LTR characters inside is not aligned/displayed properly

### ISSUE THREAD

**Issue by sixtyfive** [sqlitebrowser/sqlitebrowser#1641]
We're using (under various versions of Linux and Windows) SQLiteBrowser (currently, 3.10.1, or whatever is in distro repos) for looking at databases with lots of mixed RTL/LTR text, but even with "just" RTL text, there's a problem if it includes punctuation characters such as parentheses. C.f. below screenshot:

![image](https://user-images.githubusercontent.com/192860/49376230-e3e37800-f706-11e8-8dbd-05ea03373d66.png)

The _expected_ display would be something like this (please excuse the amateurish GIMPing):

![image](https://user-images.githubusercontent.com/192860/49376519-c95dce80-f707-11e8-9bc5-f48a114dcd1f.png)

(I'm not referring to anything I saw in #884, #695, #685, #667 or #631, but possibly #949 might be related - not sure).

**Comment by chrisjlocke** [comment]
Can you try again with the alpha release?  3.10.1 is quite old now, and we've recently released a newer alpha build.
https://github.com/sqlitebrowser/sqlitebrowser/releases

On Windows/OSX, you have the option of even newer nightly builds.

**Comment by justinclift** [comment]
Yeah.  In theory things should be at least a bit improved with the newer builds, including the recent beta1 release.  With those we use the latest Qt version (which helps), and over the last year our team as expanded to include some people more familiar with RTL languages so several bugs have been fixed.

Not sure if it's in a very good state though, but it should be better than with the v3.10.1 release. :smile:

@SafaAlfulaij Thoughts? :smile:

**Comment by SafaAlfulaij** [comment]
This happens because Qt [forces](https://bugreports.qt.io/browse/QTBUG-55012) the layout direction (LTR in this case) on any read-only text (not lineedits).
This can be solved in `ExtendedTableWidgetEditorDelegate` by:
1. Overriding [`QStyledItemDelegate::paint`](https://github.com/qt/qtbase/blob/dev/src/widgets/itemviews/qstyleditemdelegate.cpp#L379) and using the code in the bug report above
2. or (maybe) by overriding [`QStyledItemDelegate::displayText`](https://github.com/qt/qtbase/blob/dev/src/widgets/itemviews/qstyleditemdelegate.cpp#L266) and preprocessing the text to add the needed Unicode marks ([RLM, LRM, RLE, LRE](https://github.com/lxqt/pcmanfm-qt/pull/606/files#diff-a3ef47f1639253aa169b271696f97e75R468), _OR_ just prepending FSI and appending PDI to text if [Qt>=5.11](https://bugreports.qt.io/browse/QTBUG-57743)). Example: `Text` -> `(FSI)Text(PDI)`

**Comment by justinclift** [comment]
Thanks @sixtyfive & @SafaAlfulaij.  Sounds like another thing to add to our ToDo list. :smile:

**Comment by sixtyfive** [comment]
Hi, thanks all of you for looking into this! Unfortunately I don't speak C++ or else I might have taken a shot at it. I'll just stay subscribed to the issue in case anything happens.

**Comment by justinclift** [comment]
No worries @sixtyfive.  This is one of the things that's likely to improve over time, generally when someone with the right skill set comes along and feel motivated to get it done. :smile:
